### PR TITLE
ENH: Initial RTIonPlan and multiple control points support

### PR DESCRIPTION
**Pull request by MichaelColonel**
Greetings,

vtkSlicerDicomRtReader class - RTIonPlan support and control points support, some C-like arrays replaced with std::array for easier copy and assign operations. Scan spot position map inital support for modulated scan type of rt ion beam. Fraction module from DICOM is used to check if beams are available.

vtkSlicerDicomRtImportExportModuleLogic class - each control point is represented as a separate beam. Table nodes are set a children of a corresponding beam nodes.

vtkMRMLRTIonBeamNode class is added for ion beam parameters and visualization.

vtkMRMLRTBeamNode multiple visible sections of MLC opening have been added (corvus-6.2.2-phantom example data control point 10 & 11)

I can split this pull request on several parts if its very big.

### CODE REVIEW

File: Beams/MRML/vtkMRMLRTBeamNode.cxx [hunk 1]
```diff
@@ -57,6 +58,155 @@ static const char* MLCPOSITION_REFERENCE_ROLE = "MLCPositionRef";
 static const char* DRR_REFERENCE_ROLE = "DRRRef";
 static const char* CONTOUR_BEV_REFERENCE_ROLE = "contourBEVRef";
 
+//------------------------------------------------------------------------------
+namespace
+{
+
+typedef std::vector< std::pair< double, double > > PointVector;
```
Comment (cpinter) [inline Beams/MRML/vtkMRMLRTBeamNode.cxx]: Please rename variable to something more meaningful (points of what?)

File: Beams/MRML/vtkMRMLRTBeamNode.cxx [hunk 2]
```diff
@@ -57,6 +58,155 @@ static const char* MLCPOSITION_REFERENCE_ROLE = "MLCPositionRef";
 static const char* DRR_REFERENCE_ROLE = "DRRRef";
 static const char* CONTOUR_BEV_REFERENCE_ROLE = "contourBEVRef";
 
+//------------------------------------------------------------------------------
+namespace
+{
+
+typedef std::vector< std::pair< double, double > > PointVector;
+typedef std::vector< std::array< double, 4 > > LeafDataVector;
```
Comment (cpinter) [inline Beams/MRML/vtkMRMLRTBeamNode.cxx]: Same here, please don't use the word data, but call it something more informative.

Comment (cpinter) [inline Beams/MRML/vtkMRMLRTBeamNode.cxx]: Same here, please don't use the word data, but call it something more informative.

File: Beams/MRML/vtkMRMLRTBeamNode.cxx [hunk 3]
```diff
@@ -57,6 +58,155 @@ static const char* MLCPOSITION_REFERENCE_ROLE = "MLCPositionRef";
 static const char* DRR_REFERENCE_ROLE = "DRRRef";
 static const char* CONTOUR_BEV_REFERENCE_ROLE = "contourBEVRef";
 
+//------------------------------------------------------------------------------
+namespace
```
Comment (cpinter) [inline Beams/MRML/vtkMRMLRTBeamNode.cxx:63]: Why do you use a namespace here? It would be much better making this part of the class

File: Beams/MRML/vtkMRMLRTIonBeamNode.cxx [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+}
+
+//----------------------------------------------------------------------------
+vtkMRMLRTIonBeamNode::~vtkMRMLRTIonBeamNode()
+{
+  this->SetBeamDescription(nullptr);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLRTIonBeamNode::WriteXML(ostream& of, int nIndent)
```
Comment (cpinter) [inline Beams/MRML/vtkMRMLRTIonBeamNode.cxx:87]: Please use the new macros defined here
https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLNodePropertyMacros.h
See examples here
https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLDisplayNode.cxx#L110
Also for ReadXMLAttributes, Copy, and PrintSelf

File: Beams/MRML/vtkMRMLRTIonBeamNode.cxx [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+  Superclass::CreateNewBeamTransformNode();
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLRTIonBeamNode::SetVSAD( double xComponent, double yComponent)
+{
+  this->VSADx = xComponent;
+  this->VSADy = yComponent;
+  this->Modified();
+  this->InvokeCustomModifiedEvent(vtkMRMLRTBeamNode::BeamGeometryModified);
```
Comment (cpinter) [inline Beams/MRML/vtkMRMLRTIonBeamNode.cxx:242]: In some places you use Superclass, in others the class name itself. Please make it consistent.

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx [hunk 6]
```diff
@@ -77,18 +77,53 @@ class vtkSlicerDicomRtReader::vtkInternal
   public:
     BeamLimitingDeviceEntry()
       :
-      Distance(-1.),
-      NumberOfLeafJawPairs(0)
+      SourceIsoDistance(400.),
+      NumberOfPairs(0)
```
Comment (cpinter) [inline DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx]: Why did you rename this variable to have a less informative name? I'm actually curious, so if it's justifiable then no need to change, but in general we prefer longer but unambiguoug names. Not every reader of the code is an expert in every aspect of RT.

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx [hunk 7]
```diff
@@ -77,18 +77,53 @@ class vtkSlicerDicomRtReader::vtkInternal
...
+    }
+    int Number;
+    std::string ID;
+    std::string AccessoryCode;
+    std::string Type; // "ANALOG", "BINARY"
+    std::string Description;
+  };
+
+  /// Structure storing a treatment compensator parameters associated with beam
+  class CompensatorEntry
```
Comment (cpinter) [inline DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx:112]: Is this class and BlockEntry future work? Since they don't contain anything... In that case please add "//TODO: ..." comments

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx [hunk 8]
```diff
@@ -103,7 +138,7 @@ class vtkSlicerDicomRtReader::vtkInternal
     unsigned int Number;
     std::string Name;
     std::string Description;
-    double DisplayColor[3];
+    std::array< double, 3 > DisplayColor;
```
Comment (cpinter) [inline DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx:144]: I believe we used a static array because of VTK set/get macros. Is there a good reason to change this?

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx [hunk 9]
```diff
@@ -182,7 +322,7 @@ class vtkSlicerDicomRtReader::vtkInternal
   public:
     ChannelEntry()
     {
-      Number = -1;
+      Number = 0;
```
Comment (cpinter) [inline DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx]: Please explain why you changed this. -1 is an obvious "uninitialized" state, while 0 is valid, so this changes the way a newly instantiated object can be handled.

**Review by cpinter — CHANGES_REQUESTED** [review]
Happy new year :) Thanks again for the great work! I left a few comments.

Please also
- Squash these 47 commits into a single one (or 2-3 commits if they can be cleanly separated) and add descriptive comments in it, explaining in detail what changed and why (for example a line about each original commit, since you made them separately on purpose)
- Run the automated tests and make sure no new failures happen

Thank you!

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 8 above)
Comment (MichaelColonel) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx:144]: There is only one method which use DisplayColor
`double* vtkSlicerDicomRtReader::GetRoiDisplayColor(unsigned int internalIndex)`.
The return value is a pointer, so no difference. Everything else is a DCMTK specific.

No setters to change a display color, within internal elements of the reader.

I that case i had thought it's better to use std::array, rather C array (no copy, assign problems).

I can roll it back, if its required to use C array.

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 6 above)
Comment (MichaelColonel) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx]: I will return previous names.

File: Beams/MRML/vtkMRMLRTIonBeamNode.cxx (hunk 5 above)
Comment (MichaelColonel) [reply Beams/MRML/vtkMRMLRTIonBeamNode.cxx:242]: It will be Superclass only for overridden methods and inherited constructor.

File: Beams/MRML/vtkMRMLRTBeamNode.cxx (hunk 3 above)
Comment (MichaelColonel) [reply Beams/MRML/vtkMRMLRTBeamNode.cxx:63]: By means of vtkInternal, or just private?

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 7 above)
Comment (MichaelColonel) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx:112]: I will do this.

File: Beams/MRML/vtkMRMLRTIonBeamNode.cxx (hunk 4 above)
Comment (MichaelColonel) [reply Beams/MRML/vtkMRMLRTIonBeamNode.cxx:87]: I fixed that, not so sure about Copy method though.

**Comment by MichaelColonel** [comment]
Squashed to 6 commits.

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 8 above)
Comment (cpinter) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx:144]: If it's only internal then it's fine like this, thanks!

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 6 above)
Comment (cpinter) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx]: Thanks! You also changed some names to be more informative, so those should be kept (like the SourceIsoDistance)

File: Beams/MRML/vtkMRMLRTBeamNode.cxx (hunk 3 above)
Comment (cpinter) [reply Beams/MRML/vtkMRMLRTBeamNode.cxx:63]: If there is no internal class yet, then no need to bother, just private. Otherwise internal please.

**Comment by cpinter** [comment]
Thank you very much! I'll take a look shortly and integrate.
Please let me know about the automated tests.

File: DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx (hunk 6 above)
Comment (MichaelColonel) [reply DicomRtImportExport/Logic/vtkSlicerDicomRtReader.cxx]: SourceIsoDistance parameter has different physical meaning for RTPlan and RTIonPlan but the same DICOM tag (300A,00BA) 

1. For RTPlan its "Source to Beam Limiting Device Distance";
2. For RTIonPlan its "Isocenter to Beam Limiting Device Distance".

In any case the name of the parameters isn't obvious, but i will keep it.

**Comment by MichaelColonel** [comment]
23 failed out of 70

	 12 - py_DicomRtImportTest (Failed)
	 18 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseProstate_Base (Failed)
	 19 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseProstate_CERR (Failed)
	 20 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseEnt_CERR (Failed)
	 21 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseProstate_Eclipse (Failed)
	 22 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseEnt_Eclipse (Failed)
	 23 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseProstate_CERR_AutomaticOversampling (Failed)
	 24 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseEnt_CERR_AutomaticOversampling (Failed)
	 25 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseProstate_Eclipse_AutomaticOversampling (Failed)
	 26 - vtkSlicerDoseVolumeHistogramModuleLogicTest_EclipseEnt_Eclipse_AutomaticOversampling (Failed)
	 27 - vtkSlicerDoseVolumeHistogramModuleLogicTest_DoseSurfaceHistogram_EclipseEnt_Base_Inside (Failed)
	 28 - vtkSlicerDoseVolumeHistogramModuleLogicTest_DoseSurfaceHistogram_EclipseEnt_Base_Outside (Failed)
	 29 - vtkSlicerDoseVolumeHistogramModuleLogicTest_DoseSurfaceHistogram_EclipseProstate_Base_Inside (Failed)
	 30 - vtkSlicerDoseVolumeHistogramModuleLogicTest_DoseSurfaceHistogram_EclipseProstate_Base_Outside (Failed)
	 37 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Expand (Failed)
	 38 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Shrink (Failed)
	 39 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Union (Failed)
	 40 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Intersect (Failed)
	 41 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Subtract (Failed)
	 42 - vtkSlicerSegmentMorphologyModuleLogicTest_EclipseProstate_Intersect_ApplyTransform (Failed)
	 45 - vtkSlicerSegmentComparisonModuleLogicTest_EclipseProstate_Base (Failed)
	 46 - vtkSlicerSegmentComparisonModuleLogicTest_EclipseProstate_SameInput (Failed)
	 47 - vtkSlicerSegmentComparisonModuleLogicTest_EclipseProstate_Transformed (Failed)

**Comment by MichaelColonel** [comment]
Offtopic question. If i have been needed a MLC visualization like in XiO, should i create a custom node for MLC model, or i could use a beam node?

From this [post](https://github.com/SlicerRt/SlicerRT/issues/115#issuecomment-559332125)

**Comment by cpinter** [comment]
> 23 failed out of 70

Well, those will need to be fixed. I see that these failures are the same as what SlicerRT has had since the end of October. It would be great if the tests could be fixed by the developer team so that you can (after a rebase) see if your changes have broken anything or not.
I added a ticket for this (https://github.com/SlicerRt/SlicerRT/issues/125) and contacted the team about it. Please stand by for a bit until we see if you can rebase and try it again soon.

**Comment by cpinter** [comment]
About the XiO question: Is this visualization in the 3D view? If so, then special overlay would be needed to achive the same, which will require substantial work. However, you may not need it in the 3D view, in which case it is a simple slice intersection, which could already work with your changes.
Thoughts @gregsharp @lassoan ?

**Comment by MichaelColonel** [comment]
The picture (part of it) is taken from [SlicerRtData](https://github.com/SlicerRt/SlicerRtData/blob/master/xio-4.64-phantom-prostate-beams-mlc/_MLC.jpg).

**Comment by lassoan** [comment]
> About the XiO question: Is this visualization in the 3D view? If so, then special overlay would be needed to achive the same, which will require substantial work. However, you may not need it in the 3D view, in which case it is a simple slice intersection, which could already work with your changes.
Thoughts @gregsharp @lassoan ?

Since it is a beam's eye view, the camera position, orientation, and viewing angle has match the beam parameters and cannot be rotated. Since the camera cannot be rotated and viewing angle cannot be changed, it does not matter if you display the grid as a 2D or 3D actor. In 3D views it is very easy to display 3D (using model nodes), so I would just go with that. If you create a model that contains projection of the collimator leaves (their shadow) then you can use the exact same model in 2D views (as slice intersection) and in 3D view. Another advantage of this approach is that if for some reason the camera gets rotated or any way misaligned with the beam's eye view, it will still be correct (but you'll see 3D beams and not just a 2D grid). In beam's eye 3D views you can choose to disable camera rotation similarly to how it is done for slice views [here](https://www.slicer.org/wiki/Documentation/Nightly/ScriptRepository#Disable_certain_user_interactions_in_slice_views).

**Comment by cpinter** [comment]
@MichaelColonel The problem that made most of the tests fail has been fixed by @Sunderlandkyl, and the dashboard shows only one failing test. So I think if you can build this branch against the latest Slicer and only the same tests fails (py_DicomRtImportTest, which tests importing/loading DICOM but not ion plan obviously - it may be added to it though), then I'll integrate this. Thanks again!

**Comment by MichaelColonel** [comment]
All tests have passed except py_DicomRtImportTest. It failed because number of loaded beam nodes, and total number of nodes in hierarchy had increased (each control point is a separated beam node).

After some minor changes:

1.   `self.assertEqual( len( slicer.util.getNodes('vtkMRMLModelHierarchyNode*') ), 11 )`
2.   `self.assertEqual( shNode.GetNumberOfItems(), 42 )`

it also passed.

These changes are not in pull request.

**Comment by cpinter** [comment]
@MichaelColonel you are great :) Now that the tests pass, I'll integrate this change, but not before you give me the green light, because I have seen pushes recently, and not sure if you want to add more improvements.

I'll make the changes in py_DicomRtImportTest and close the related ticket. Thanks again!

**Comment by MichaelColonel** [comment]
The pull request is ready to be merged, latest pushes were minor API style changes - GetBeamControlPoint* instead of GetControlPoint*.
